cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillPreparer.java

```diff
@@ -385,46 +397,85 @@ private LocalFile createLocalFile(String name, Location location, boolean archiv
 
-      List<Class<?>> classes = Lists.newArrayList();
-      classes.add(ApplicationMasterMain.class);
+      Location cachedLocation = jarCacheLocationFactory == null ?
+        null : jarCacheLocationFactory.create(Constants.Files.APP_MASTER_JAR);
+      if (cachedLocation != null && cachedLocation.exists()) {
+        LOG.debug("Found cached app master jar for twill app {} at {}", twillSpec.getName(), cachedLocation);
+        // the jar is cached on local disk. Upload it to hdfs.
+        copyContents(cachedLocation, location);
+      } else {
+        List<Class<?>> classes = Lists.newArrayList();
+        classes.add(ApplicationMasterMain.class);
+
+        // Stuck in the yarnAppClient class to make bundler being able to pickup the right yarn-client version
+        classes.add(yarnAppClient.getClass());
+
+        // Add the TwillRunnableEventHandler class
+        if (twillSpec.getEventHandler() != null) {
+          classes.add(getClassLoader().loadClass(twillSpec.getEventHandler().getClassName()));
+        }
 
-      // Stuck in the yarnAppClient class to make bundler being able to pickup the right yarn-client version
-      classes.add(yarnAppClient.getClass());
+        bundler.createBundle(location, classes);
+        LOG.debug("Done {}", Constants.Files.APP_MASTER_JAR);
 
-      // Add the TwillRunnableEventHandler class
-      if (twillSpec.getEventHandler() != null) {
-        classes.add(getClassLoader().loadClass(twillSpec.getEventHandler().getClassName()));
+        if (cachedLocation != null) {
+          // to handle race conditions, only the first one to create the cached location will populate it
+          if (cachedLocation.createNew()) {
+            copyContents(location, cachedLocation);
+          }
+        }
       }
-
-      bundler.createBundle(location, classes);
-      LOG.debug("Done {}", Constants.Files.APP_MASTER_JAR);
-
       localFiles.put(Constants.Files.APP_MASTER_JAR, createLocalFile(Constants.Files.APP_MASTER_JAR, location));
+
     } catch (ClassNotFoundException e) {
       throw Throwables.propagate(e);
     }
   }
 
   private void createContainerJar(ApplicationBundler bundler, Map<String, LocalFile> localFiles) throws IOException {
     try {
-      Set<Class<?>> classes = Sets.newIdentityHashSet();
-      classes.add(TwillContainerMain.class);
-      classes.addAll(dependencies);
-
-      ClassLoader classLoader = getClassLoader();
-      for (RuntimeSpecification spec : twillSpec.getRunnables().values()) {
-        classes.add(classLoader.loadClass(spec.getRunnableSpecification().getClassName()));
-      }
-
       LOG.debug("Create and copy {}", Constants.Files.CONTAINER_JAR);
       Location location = createTempLocation(Constants.Files.CONTAINER_JAR);
-      bundler.createBundle(location, classes, resources);
-      LOG.debug("Done {}", Constants.Files.CONTAINER_JAR);
+
+      Location cachedLocation = jarCacheLocationFactory == null ?
```

Comment:
these jars aren't different per application. They are the same for each program type.